### PR TITLE
Transpose optimization for Softmax and LogSoftmax (fixes #1716)

### LINKED ISSUE #1716: Feature: Transpose optimization for Softmax
Thanks to the optimizations by tf2onnx, most Transpose operations in converted models are eliminated.
However, Softmax does not seem to be accounted for yet.

Take the following simple SavedModel:
```py
import tensorflow as tf
from keras import Model
from keras.layers import Input, Softmax
import keras.backend as K
K.set_image_data_format('channels_first')

inputs = Input((1, None, None), name='input')
outputs = Softmax(axis=1, name='output')(inputs)
m = Model(inputs, outputs)

tf.saved_model.save(
    obj=m,
    export_dir='test_softmax'
)
```

When exporting it via `python -m tf2onnx.convert --saved-model test_softmax --output test_softmax/model.onnx --tag serve --signature_def serving_default --opset 14`, the result looks like this:

Same for every other channels_first model I export - the Softmax operation is always surrounded by Transposes.

### PR DESCRIPTION
**Pull request by janbernloehr**
This is reopening #1727 by @fthielke to fix the remaining issues to get it merged.

**Original Description**
In opsets 13 and higher, the axis of the operation is arbitrary and can simply be changed according to the permutation of the Transpose.
In lower opsets, Softmax always coerces its inputs to a 2D tensor, making Transpose operations necessary if the permutation moves axes between the coerced batch and feature dimensions.

### CODE REVIEW

**Comment by janbernloehr** [comment]
@fatcat-z , @TomWildenhain-Microsoft : Maybe you can take another look. pylint issues have been fixed.

**Review by fatcat-z — APPROVED** [review]
LGTM, thanks for your contributions!
